File: ios/ReelVaultCore/Reel.swift
import Foundation

public struct ContentStep: Codable, Hashable, Sendable {
    public let type: String
    public let description: String

    public init(type: String, description: String) {
        self.type = type
        self.description = description
    }
}

public struct Reel: Identifiable, Codable, Sendable {
    public let id: String
    public let userId: String?
    public let sourcePlatform: String?
    public let originalUrl: String
    public let shortcode: String?
    public let creatorUsername: String?
    public let creatorName: String?
    public let title: String?
    public let topic: String?
    public let summary: String?
    public let hook: String?
    public let transcript: String?
    public let cleanedTranscript: String?
    public let contentStructure: [ContentStep]?
    public let tags: [String]?
    public let thumbnailUrl: String?
    public let durationSeconds: Int?
    public let status: ProcessingStatus
    public let errorMessage: String?
    public let createdAt: String
    public let processedAt: String?
    public let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case sourcePlatform = "source_platform"
        case originalUrl = "original_url"
        case shortcode
        case creatorUsername = "creator_username"
        case creatorName = "creator_name"
        case title
        case topic
        case summary
        case hook
        case transcript
        case cleanedTranscript = "cleaned_transcript"
        case contentStructure = "content_structure"
        case tags
        case thumbnailUrl = "thumbnail_url"
        case durationSeconds = "duration_seconds"
        case status
        case errorMessage = "error_message"
        case createdAt = "created_at"
        case processedAt = "processed_at"
        case updatedAt = "updated_at"
    }
}

public struct ReelListDTO: Codable, Sendable {
    public let data: [Reel]
    public let total: Int
    public let page: Int
    public let limit: Int
}

File: backend/app/schemas/reel.py
import re
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field, field_validator
from app.models.enums import ProcessingStatus


class ReelCreate(BaseModel):
    url: str = Field(..., description="Instagram Reel URL (e.g. https://www.instagram.com/reel/C8abc123/)")

    @field_validator("url")
    @classmethod
    def validate_url(cls, v: str) -> str:
        clean_url = v.strip()
        # Basic check for instagram URL
        pattern = r"instagram\.com/(reel|reels|p)/([A-Za-z0-9_-]+)"
        if not re.search(pattern, clean_url, re.IGNORECASE):
            raise ValueError("URL must be a valid Instagram Reel or Post link (e.g. https://www.instagram.com/reel/ABC123/)")
        return clean_url


class ReelResponse(BaseModel):
    id: str
    status: ProcessingStatus
    original_url: str
    shortcode: Optional[str] = None
    created_at: datetime
    message: Optional[str] = "Reel queued for processing."


class ContentStructureItem(BaseModel):
    type: str
    description: str


class ReelDetail(BaseModel):
    id: str
    user_id: Optional[str] = None
    source_platform: str = "instagram"
    original_url: str
    shortcode: Optional[str] = None
    creator_username: Optional[str] = None
    creator_name: Optional[str] = None
    title: Optional[str] = None
    topic: Optional[str] = None
    summary: Optional[str] = None
    hook: Optional[str] = None
    transcript: Optional[str] = None
    cleaned_transcript: Optional[str] = None
    content_structure: Optional[List[Dict[str, Any]]] = None
    tags: List[str] = Field(default_factory=list)
    thumbnail_url: Optional[str] = None
    duration_seconds: Optional[int] = None
    status: ProcessingStatus
    error_message: Optional[str] = None
    created_at: datetime
    processed_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ReelListResponse(BaseModel):
    data: List[ReelDetail]
    total: int
    page: int
    limit: int
